App/BraveWidgets/News Topics/NewsTopicsModel.swift

```diff
@@ -0,0 +1,163 @@
+// Copyright 2023 The Brave Authors. All rights reserved.
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+import Foundation
+import CodableHelpers
+import Collections
+import UIKit
+
+/// Handles fetching Brave News Topics for widgets
+struct NewsTopicsModel {
+  var fetchNewsTopics: @Sendable () async -> [NewsTopic]
+  var fetchImageThumbnailsForTopics: @Sendable ([NewsTopic], CGSize) async -> [NewsTopic.ID: UIImage]
+}
+
+extension NewsTopicsModel {
+  /// The live implementation of the news topic model
+  ///
+  /// - note: For now this implementation does not cache the results of the file, since the widget is not
+  /// reloaded by the app in any way yet. Hopefully we can move this code to the BraveNews module in the
+  /// future without bloating the size of the widget extension and then get all the cacheing mechanisms
+  /// from there.
+  static var live: Self {
+    .init(
+      fetchNewsTopics: {
+        do {
+          // At the moment there is only english US topics
+          let url = URL(string: "https://brave-today-cdn.brave.com/news-topic-clustering/widget_topic_news.en_US.json")!
+          let session = URLSession(configuration: .default)
+          let (data, _) = try await session.data(for: URLRequest(url: url))
+          
+          let topics = Dictionary(
+            grouping: try JSONDecoder.topicsDecoder.decode([FailableDecodable<NewsTopic>].self, from: data)
+              .compactMap(\.wrappedValue).sorted(by: >),
+            by: \.topicIndex
+          )
+          let maxCount = topics.values.map(\.count).max() ?? 0
+          var articles: OrderedSet<NewsTopic> = .init()
+          for i in 0..<maxCount {
+            for key in topics.keys.sorted() {
+              if let article = topics[key]?[safe: i] {
+                articles.append(article)
+              }
+            }
+          }
+          return Array(articles)
+        } catch {
+          return []
+        }
+      },
+      fetchImageThumbnailsForTopics: { topics, thumbnailSize in
+        return await withTaskGroup(of: (String, UIImage?).self, returning: [String: UIImage].self) { group in
+          var images: [String: UIImage] = [:]
+          let session = URLSession(configuration: .default, delegate: nil, delegateQueue: {
+            let queue = OperationQueue()
+            queue.maxConcurrentOperationCount = 3
+            return queue
+          }())
+          for topic in topics {
+            guard let imageURL = topic.imageURL else { continue }
+            group.addTask {
+              do {
+                let request = URLRequest(url: imageURL, timeoutInterval: 10)
```

Comment:
Is this timeout too low for slow networks?

Reply:
Its a good question, at most we're downloading 15 images at 3 at a time, so 50s on the large lists. We dont want to have the timeout that long or the widget takes longer to show. We can maybe bump this up a bit but Im not sure what the highest we'd want to go is.